### PR TITLE
Bezier Base

### PR DESCRIPTION
**Pull request by gaganjyot**

### CODE REVIEW

File: lckernel/cad/geometry/geobezier.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    return Coordinate(x_, y_);
+}
+
+const std::vector<Coordinate> Bezier::Curve() {
+    std::vector<Coordinate> v = { _pointA, _pointB, _pointC };
+    std::vector<Coordinate> ret;
+    for(int i = 0.; i < 1.; i+=.001) {
+        ret.push_back(CasteljauAt(v, i));
+    }
+    const std::vector<Coordinate> const_ret = std::move(ret);
```
Comment (dxli) [inline lckernel/cad/geometry/geobezier.cpp]: this seems to be weird to me.

Why doesn't do the straight forward:

```C++
const std::vector<Coordinate> Bezier::Curve() const {
    std::vector<Coordinate> v = {{ _pointA, _pointB, _pointC }};
     std::vector<Coordinate> ret;
     for(int i = 0.; i < 1.; i+=.001) {
        ret.push_back(CasteljauAt(v, i));
     }
return ret;
}
```

File: lckernel/cad/geometry/geobezier.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+}
+
+Coordinate Bezier::DirectValueAt(float t) const {
+    double one_minus_t = 1 - t;
+    double t_square = t*t;
+    double one_minus_t_square = one_minus_t * one_minus_t;
+    double two_a_b  = 2 * one_minus_t * t;
+    double x_ = (one_minus_t_square * _pointA.x()) + (two_a_b * _pointB.x()) + (t_square * _pointC.x());
+    double y_ = (one_minus_t_square * _pointA.y()) + (two_a_b * _pointB.y()) + (t_square * _pointC.y());
```
Comment (dxli) [inline lckernel/cad/geometry/geobezier.cpp]: why not do the calculation in Coordinates:

```C++
return _pointA * one_minus_t_square +   _pointB * two_a_b +  _pointC * t_square;

File: lckernel/cad/geometry/geobezier.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if(points.size() == 1) return points[0];
+    else {
+        int size_ = points.size() - 1   ;
+        std::vector<Coordinate> new_vec;
+        for(int i = 0; i < size_; i++) {
+            double _x = (1-t) * points[i].x() + t * points[i+1].x();
+            double _y = (1-t) * points[i].y() + t * points[i+1].y();
+            new_vec.push_back(Coordinate(_x, _y));
+        }
+        return CasteljauAt(new_vec, t);
```
Comment (dxli) [inline lckernel/cad/geometry/geobezier.cpp]: trying to make it compact here. Should be easier for maintenance

```C++
Coordinate Bezier::CasteljauAt(std::vector<Coordinate> points, float t) const {
 if(points.size() == 1) return points[0];
std::vector<Coordinate> new_vec;
for (size_t i = 1; i < points.size(); i++)
     new_vec.push_back( points[i-1] * (1 - t) + points[i] * t);
return new_vec;
}
```

Anyway, most time, we should be using matrix based algorithms for efficiency.



}
```

File: lckernel/cad/geometry/geobezier.cpp (hunk 1 above)
Comment (gaganjyot) [inline lckernel/cad/geometry/geobezier.cpp]: Agreed, I was thinking of hardcoded precission too. regarding std::move, I was playing with move and understanding it. It was not supposed to be there. will take care of it.

File: lckernel/cad/geometry/geobezier.cpp (hunk 2 above)
Comment (gaganjyot) [inline lckernel/cad/geometry/geobezier.cpp]: makes better sense now. Thanks :+1:

File: lckernel/cad/geometry/geobezier.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    double t_square = t*t;
+    double one_minus_t_square = one_minus_t * one_minus_t;
+    double two_a_b  = 2 * one_minus_t * t;
+    return (_pointA * one_minus_t_square) + (_pointB * two_a_b) + (_pointC * t_square);
+}
+
+const std::vector<Coordinate> Bezier::Curve(float precession) {
+    std::vector<Coordinate> v = { _pointA, _pointB, _pointC };
+    std::vector<Coordinate> ret;
+    for(int i = 0.; i < 1.; i+=precession) {
```
Comment (dxli) [inline lckernel/cad/geometry/geobezier.cpp]: i should be double instead of integer

**Pull request merged**